hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/request/EdgeRequestBuilder.java

```diff
@@ -127,12 +111,40 @@ private Single<EdgeSetGroupRequest> buildEdgeRequest(
                             timeRange,
                             space,
                             neighborIds,
-                            edgesByType.get(entityType)),
-                filters));
+                            edgesSelectionsByType.get(entityType))));
+  }
+
+  private Single<Map<String, EdgeSetRequest>> getEntityTypeToEdgeSetRequest(
+      GraphQlRequestContext context, EdgeType edgeType, Set<SelectedField> edgeFields) {
+    return Observable.fromIterable(edgeFields)
+        .collect(Collectors.groupingBy(this::getEntityType, Collectors.toUnmodifiableSet()))
+        .flatMap(entry -> this.getEdgeSetRequest(context, edgeType, entry));
   }
 
-  private Map<String, Set<SelectedField>> getEdgesByType(Stream<SelectedField> edgeSetStream) {
+  private Single<Map<String, EdgeSetRequest>> getEdgeSetRequest(
+      GraphQlRequestContext context,
+      EdgeType edgeType,
+      Map<String, Set<SelectedField>> edgeFields) {
+    return Observable.fromIterable(edgeFields.entrySet())
+        .flatMapSingle(
+            entry ->
+                this.getEdgeSetRequestEntry(context, edgeType, entry.getKey(), entry.getValue()))
+        .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue));
+  }
 
+  private Single<Map.Entry<String, EdgeSetRequest>> getEdgeSetRequestEntry(
+      GraphQlRequestContext context, EdgeType edgeType, String key, Set<SelectedField> edgeFields) {
```

Comment:
edgeFields here really means all edge set fields for the given key (where key is really entity type) and edgeType.